### ISSUE TITLE
False positive for redundant_closure_call when called in a closure

### ISSUE THREAD

**Issue by Systemcluster** [rust-lang/rust-clippy#5916]
Minimal example:

```rust
fn fun<T: 'static + FnMut()>(mut f: T) {
    f();
}

fn main() {
    let mut x;
    let create = || 2 * 2;
    x = create();

    fun(move || {
        println!("using x here {}", x);
        x = create();
        println!("to avoid non-relevant warnings {}", x);
    })
}
```

`create` is called in two spots, but on the first `x = create();` a `warning: closure called just once immediately after it was declared` is emitted.

Interestingly, the warning is not emitted when any statement is between the definitian and the first usage of the closure:
```rust
let create = || 2 * 2;
println!("no warning emitted with this line");
x = create();
```

### Meta

- `cargo clippy -V`: `clippy 0.0.212 (9b88e0a86 2020-08-15)`
- `rustc -Vv`:
    ```
    rustc 1.47.0-nightly (9b88e0a86 2020-08-15)
    binary: rustc
    commit-hash: 9b88e0a8667040452a94fda8548f0f5f0f801f90
    commit-date: 2020-08-15
    host: x86_64-pc-windows-msvc
    release: 1.47.0-nightly
    LLVM version: 10.0
    ```

**Comment by flip1995** [comment]
This lint is probably implemented with a visitor. The fix is easy: The [`nested_visit_map`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/trait.Visitor.html#tymethod.nested_visit_map) method should at least return [`NestedVisitorMap::OnlyBodies`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/enum.NestedVisitorMap.html#variant.OnlyBodies)

**Comment by giraffate** [comment]
I will prepare a fix.
